### PR TITLE
Improve ETagRequired error range

### LINKED ISSUE #876: Improve ETagRequired error range
Given this XML file:

```xml
<foo>
  <bar></
</foo>
```

We have that:

image

Although the ETagRequired error range should highlight the start tag element:

### PR DESCRIPTION
**Pull request by angelozerr**
Improve ETagRequired error range

Fixes #876

Signed-off-by: azerr azerr@redhat.com

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR fixes error range for ETAgrequired, the main idea is to higlight the error from the start tag which is broken to the offset reported by Xerces:

image

The start tag can be broken because:

 * it has not an end tag
 * it has an end tag but it contains </ which breaks the XML.

As I had to fix CloseStartTagAction which is used for several error code,this PR fixes several issues:

 * https://github.com/eclipse/lemminx/issues/900
 * https://github.com/eclipse/lemminx/issues/889
 * https://github.com/eclipse/lemminx/issues/880
 * https://github.com/eclipse/lemminx/issues/878

**Comment by datho7561** [comment]
For cases like this:
```xml
<a>
  <c><b>
</a>
```
the opening `<` of `</a>` is highlighted in the error range:

Screenshot from 2020-10-05 10-01-02

I find this behaviour odd, since the closing tag should appear before `</a>`. As well, `</a>` is correct; it closes the `<a>` tag.

**Review by datho7561 — CHANGES_REQUESTED** [review]

**Comment by angelozerr** [comment]
> the opening < of </a> is highlighted in the error range:

fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/XMLSyntaxErrorCode.java [hunk 1]
```diff
@@ -216,7 +218,28 @@ public static Range toLSPRange(XMLLocator location, XMLSyntaxErrorCode code, Obj
...
+					tagElement = findTag(tag, element);
+
+				}
+
+			}
+			if (tagElement != null) {
+				return XMLPositionUtility.createRange(tagElement.getStartTagOpenOffset(), offset, document);
+			}
+			// String tag = getString(arguments[0]);
+			// return XMLPositionUtility.selectChildEndTag(tag, offset, document);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/XMLSyntaxErrorCode.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/ETagUnterminatedCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/*******************************************************************************
+* Copyright (c) 2019 Red Hat Inc. and others.
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/ETagUnterminatedCodeAction.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+
+/**
+ * Code action to fix close start tag element.
+ *
+ */
+public class CloseTagCodeAction implements ICodeActionParticipant {
+
+	public void doCodeAction(Diagnostic diagnostic, Range range, DOMDocument document, List<CodeAction> codeActions,
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java:37]: fixed and I add comments with samples for each usecases.

**Comment by angelozerr** [comment]
@datho7561 please review my PR (with vscode-xml and see if my code is understandable). Thanks

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java:37]: This method is still pretty long. Do you think its possible to split the cases for `startTagClosed == true` and `startTagClosed == false` into separate functions?

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java:37]: Try to fixed that, is it better?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+			} else {
+				// Here start tag element is closed with '>'.
+				if (!element.hasEndTag()) {
+					// The element has no an end tag
+					// ex : <foo attr="" >
+					// // Close with '</foo>'
+					String tagName = element.getTagName();
+					String label = "</" + tagName + ">";
+					final String initialInsertText = label;
+					String insertText = initialInsertText;
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/CloseTagCodeAction.java]: fixed

**Pull request merged**